### PR TITLE
[feature](Nereids): Infer isNotNull from filter and eliminate OuterJoin

### LINKED ISSUE #16707: outer join elimination

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #16707 

## Problem summary

Describe your changes.

- current infer `infer not null slot` code exist bug.
- add `inferIsNotNull` and modify `EliminateOuterJoin` which will use `is not null` to eliminate outer join.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.7 seconds
 load time: 482 seconds
 storage size: 17171809400 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230203151522_clickbench_pr_90296.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Expression.java [hunk 1]
```diff
@@ -45,8 +45,8 @@
  * Abstract class for all Expression in Nereids.
  */
 public abstract class Expression extends AbstractTreeNode<Expression> implements ExpressionTrait {
-
-    private static final String INPUT_CHECK_ERROR_MESSAGE = "argument %d requires %s type, however '%s' is of %s type";
+    // Mask this expression is generated by rule, should be removed.
+    public boolean isGeneratedIsNotNull = false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Expression.java:49]: should we process this attr in each withXXX function? i think we will lose this tag after rewrite.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateNotNull.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                Set<Expression> keepIsNotNull = slotsFromIsNotNull.stream()
+                        .filter(ExpressionTrait::nullable)
+                        .filter(slot -> !inferNonNotSlots.contains(slot))
+                        .map(slot -> new Not(new IsNull(slot))).collect(Collectors.toSet());
+
+                // merge removeIsNotNullPredicates and keepIsNotNull into a new ImmutableSet
+                Set<Expression> newPredicates = ImmutableSet.<Expression>builder()
+                        .addAll(removeIsNotNullPredicates)
+                        .addAll(keepIsNotNull)
+                        .build();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateNotNull.java:78]: this rule is hard to understand, please add some example to explain it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 3]
```diff
@@ -101,6 +101,9 @@ public enum RuleType {
     ELIMINATE_GROUP_BY_CONSTANT(RuleTypeClass.REWRITE),
     ELIMINATE_ORDER_BY_CONSTANT(RuleTypeClass.REWRITE),
     INFER_PREDICATES(RuleTypeClass.REWRITE),
+    INFER_FILTER_NOT_NULL(RuleTypeClass.REWRITE),
+    INFER_JOIN_NOT_NULL(RuleTypeClass.REWRITE),
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:105]: next PR will use it.

**Review by morrySnow — APPROVED** [review]

**Pull request merged**